pkg/i2gw/aggregator_test.go

```diff
@@ -0,0 +1,358 @@
+					ParentRefs: []gatewayv1beta1.ParentReference{{
+						Name: "example-proxy",
+					}},
+				},
+				Hostnames: []gatewayv1beta1.Hostname{"example.net"},
+				Rules: []gatewayv1beta1.HTTPRouteRule{{
+					Matches: []gatewayv1beta1.HTTPRouteMatch{{
+						Path: &gatewayv1beta1.HTTPPathMatch{
+							Type:  &gExact,
+							Value: stringPtr("/bar"),
+						},
+					}},
+					BackendRefs: []gatewayv1beta1.HTTPBackendRef{{
+						BackendRef: gatewayv1beta1.BackendRef{
+							BackendObjectReference: gatewayv1beta1.BackendObjectReference{
+								Name:  "custom",
+								Group: apiGroupPtr("vendor.example.com"),
+								Kind:  apiKindPtr("StorageBucket"),
+							},
+						},
+					}},
+				}},
+			},
+		}, {
+			ObjectMeta: metav1.ObjectMeta{Name: "net-default-backend", Namespace: "different"},
+			Spec: gatewayv1beta1.HTTPRouteSpec{
+				CommonRouteSpec: gatewayv1beta1.CommonRouteSpec{
+					ParentRefs: []gatewayv1beta1.ParentReference{{
+						Name: "example-proxy",
+					}},
+				},
+				Rules: []gatewayv1beta1.HTTPRouteRule{{
+					BackendRefs: []gatewayv1beta1.HTTPBackendRef{{
+						BackendRef: gatewayv1beta1.BackendRef{
+							BackendObjectReference: gatewayv1beta1.BackendObjectReference{
+								Name: "default",
+								Port: portNumberPtr(8080),
+							},
+						}},
+					}},
+				},
+			},
+		}},
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			aggregator := ingressAggregator{ruleGroups: map[ruleGroupKey]*ingressRuleGroup{}}
+
+			for _, ingress := range tc.ingresses {
+				aggregator.addIngress(ingress)
+			}
+
+			httpRoutes, gateways, errors := aggregator.toHTTPRoutesAndGateways()
+
+			if len(httpRoutes) != len(tc.expectHttpRoutes) {
+				t.Errorf("Expected %d HTTPRoutes, got %d: %+v", len(tc.expectHttpRoutes), len(httpRoutes), httpRoutes)
+			} else {
+				for i, got := range httpRoutes {
+					want := tc.expectHttpRoutes[i]
+					want.SetGroupVersionKind(httpRouteGVK)
+					if !apiequality.Semantic.DeepEqual(got, want) {
+						t.Errorf("Expected HTTPRoute %d to be %+v\n Got: %+v\n Diff: %s", i, want, got, cmp.Diff(want, got))
+					}
+				}
+			}
+
+			if len(gateways) != len(tc.expectGateways) {
+				t.Errorf("Expected %d Gateways, got %d: %+v", len(tc.expectGateways), len(gateways), gateways)
+			} else {
+				for i, got := range gateways {
+					want := tc.expectGateways[i]
+					want.SetGroupVersionKind(gatewayGVK)
+					if !apiequality.Semantic.DeepEqual(got, want) {
+						t.Errorf("Expected Gateway %d to be %+v\n Got: %+v\n Diff: %s", i, want, got, cmp.Diff(want, got))
+					}
+				}
+			}
+
+			if len(errors) != len(tc.expectErrors) {
+				t.Errorf("Expected %d errors, got %d: %+v", len(tc.expectErrors), len(errors), errors)
+			} else {
+				for i, e := range errors {
+					if e != tc.expectErrors[i] {
+						t.Errorf("Unexpected error message at %d index. Got %s, want: %s", i, e, tc.expectErrors[i])
+					}
+				}
+			}
+		})
+	}
+}
+
+func stringPtr(s string) *string {
```

Comment:
we have utils for this in https://pkg.go.dev/k8s.io/utils/pointer, but that doesn't need to hold anything up, just a heads up.